android/samples/gltf-bloom/app/src/main/java/com/google/android/filament/gltf/MainActivity.kt

```diff
@@ -0,0 +1,587 @@
+
+import com.google.android.filament.gltfio.AssetLoader
+import com.google.android.filament.gltfio.FilamentAsset
+import com.google.android.filament.gltfio.MaterialProvider
+import com.google.android.filament.gltfio.ResourceLoader
+
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.nio.channels.Channels
+import kotlin.math.*
+
+class Framebuffer {
+    var camera: Camera? = null
+    var view: View? = null
+    var scene: Scene? = null
+    var color: Texture? = null
+    var depth: Texture? = null
+    var target: RenderTarget? = null
+}
+
+class MainActivity : Activity() {
+
+    // Be sure to initialize not only Filament, but also gltfio (via AssetLoader)
+    companion object {
+        init {
+            Filament.init()
+            AssetLoader.init()
+        }
+    }
+
+    private lateinit var surfaceView: SurfaceView
+    private lateinit var uiHelper: UiHelper
+    private lateinit var choreographer: Choreographer
+
+    private lateinit var engine: Engine
+    private lateinit var renderer: Renderer
+
+    private lateinit var assetLoader: AssetLoader
+    private lateinit var filamentAsset: FilamentAsset
+
+    private lateinit var finalScene: Scene
+    private lateinit var finalView: View
+    private lateinit var finalCamera: Camera
+
+    private val kGaussianSampleCount = 9
+    private val kGaussianFilterSize = 1 + 2 * (kGaussianSampleCount - 1)
+    private val kGaussianWeights = FloatArray(kGaussianSampleCount * 4 * 2)
+
+    @Entity private var finalQuad = 0
+    @Entity private var hblurQuad = 0
+    @Entity private var vblurQuad = 0
+
+    private val primary = Framebuffer()
+    private val hblur = Framebuffer()
+    private val vblur = Framebuffer()
+
+    private var quadVertBufferGpu: VertexBuffer? = null
+    private var quadIndxBufferGpu: IndexBuffer? = null
+
+    enum class ImageOp { MIX, HBLUR, VBLUR}
+
+    private lateinit var mixMaterial: Material
+    private lateinit var blurMaterial: Material
+
+    private lateinit var ibl: Ibl
+    @Entity private var light = 0
+
+    private var swapChain: SwapChain? = null
+
+    private val frameScheduler = FrameCallback()
+
+    private val animator = ValueAnimator.ofFloat(0.0f, (2.0 * PI).toFloat())
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        computeGaussianWeights()
+
+        surfaceView = SurfaceView(this)
+        setContentView(surfaceView)
+
+        choreographer = Choreographer.getInstance()
+
+        uiHelper = UiHelper(UiHelper.ContextErrorPolicy.DONT_CHECK)
+        uiHelper.renderCallback = SurfaceCallback()
+        uiHelper.setDesiredSize(720, 1280)
+        uiHelper.attachTo(surfaceView)
+
+        engine = Engine.create()
+        renderer = engine.createRenderer()
+
+        primary.scene = engine.createScene()
+        primary.camera = engine.createCamera()
+        primary.view = engine.createView()
+        primary.view!!.setName("primary")
```

Comment:
Instead of using `!!` everywhere you should make `Framebuffer` be a `data class` with non-nullable fields and create an instance like so:

`primary = Framebuffer(engine.createScene(), engine.createCamera(), engine.createView(), …)`

Reply:
Done.